Documentation/facet-configurations.md

```diff
@@ -0,0 +1,278 @@
+# Facet validity
+
+This working document defines the possible configurations and intents for the various facets and their configurations.
+
+## Entities
+
+IDS Wall - SubType Marco
+IDS Wall - SubType USERDEFINED
+
+WALL - PARAPET
+WALL - USERDEFINED -> Claudio's Type
+WALL - USERDEFINED -> Marco's Type
+
+NAME is mandatory.
+
+| Fields Entered   | Required | Optional | Prohibited | Applicability |
+| ---------------- | -------- | -------- | ---------- | ------------- |
+| NAME             | ✅       | ❌       | ❌         | ✅            |
+| NAME / [SUBTYPE] | ✅       | ❌       | ❌         | ✅            | SCHEMA CHANGE!!! (NAME)
+
+Optional Entity requirements have no meaning? An Applicable Element is a IFCDOOR or it is not.
+
+Always express entity in positive terms.
+Disallowed IFC types can be expressed prohibiting their applicability.
+
+The logic for the identification of values for SUBTYPE is to return an enumeration of values:
+
+1. If a type is defined via the IfcRelDefinesByType relation
+  the PredefinedType of the type overrides the entity's PredefinedType (which should be empty according to IFC documentation)
+  1.1 if Type's PredefinedType is .USERDEFINED.
+    -> IfcElementType.ElementType
+    -> .USERDEFINED.
+  1.2 else Type's PredefinedType (if not null)
+    -> Type's - PredefinedType
+2. the value of the direct attribute PredefinedType
+  2.1 if the attribute PredefinedType == .USERDEFINED. then the value of the attribute IfcElement/ObjectType is used
+  2.2 else entity's PredefinedType (if not null)
+3. TODO document the same for ElementType and ProcessType
+
+TODO: check if the validation service of IFC prevents entity level types to be defined when type's type is specified.
```

Comment:
The rationale for this change to the way entity facet's PredefinedType behaviour makes total sense. However it may be worth considering the implications for implementors (both during spec creation and model validation). Particularly for 'early-bound' IFC implementations I can see challenges when traversing from an Element (generally of a known type) to any defining ElementType's PredefinedType enum attribute. The issue being you don't know what Type you'll get at runtime, which provides the enum Type.

Take as an example: an IDS specifier can presumably define an entity facet as:
**Name**: IFCFLOWTERMINAL
**SubType**: BATH  _(the enum from IfcSanitaryTerminalTypeEnum)_

(Obviously in IFC4 onward you might be able to use **Name** = IfcSanitaryTerminal, but not in 2x3)

So two issues:
1. The design interface (and IDS validator) would need to prompt/allow all appropriate enum values for a FlowTerminal -  from all potential DefiningTypes
2. In order to query that PredefinedType value on the ElementType at runtime you've some casting and coercion to do, and a runtime performance hit.

All doable, but it would be a lot more efficient and easier to implement if there were some means to specify the Type as part of the entity. It would also be more expressive: you could select all sanitary equipment in IFC2x3.

I guess the question is: are we considering these kinds of complex uses of PredefinedType/SubType for IDS 1.0? Entity Facets seem most often used on _applicability_ rather than _requirements_ so it's important these kinds of query predicates can be employed consistently across implementations or we'll be testing different sets of applicable items.